### DISCUSSION TITLE
Resources with step-by-step example calculations to better understand CALPHAD?

### DISCUSSION THREAD

**Discussion by Jarrod-Angove** [pycalphad/pycalphad#553] [Q&A] [answered]
I've been trying to gain a better understanding of the inner workings of the CALPHAD method. I've gone through the basics in two textbooks and have completed a graduate level materials thermodynamics class, but I still feel that I don't fully understand what is going on behind the abstractions provided by pycalphad. Ideally, there's some resource/guide out there that goes through some of the fundamental calculations used in the CALPHAD method in a concrete way. This would involve starting with the information available in a TDB file, selecting the relevant models for the chosen problem, and building up the optimization problem by hand. Even better would be an example where the problem isn't trivial (i.e. predicting a binary phase diagram for an alloy with >4 components). I don't think it's necessary (or reasonable) to demonstrate the optimization, but getting to the point where the objective function is defined would be amazing. 

Does a resource like this exist? Am I better off just reading the pycalphad source code? Any help would be hugely appreciated!

Thanks :)

**Comment by bocklund** [comment, marked as answer]
Thanks, this is a really good and important question.

A singular resource that covers everything in terms of the PyCalphad abstractions doesn't exist right now. To my knowledge, there's also no self-contained software-independent resource that covers all everything succinctly and completely, including the numerical methods one needs to implement a Calphad-type code. The textbooks that I'm familiar with tend to be aimed at practitioners, model developers, and database maintainers rather than software developers. Most of the job of the software tools is to abstract away the numerical methods.

I made some slides a few years ago that might be insightful to some of the PyCalphad specifics: [2021-02-12-Bocklund-PRL-group-seminar.pdf](https://github.com/user-attachments/files/16398285/2021-02-12-Bocklund-PRL-group-seminar.pdf). A lot of the fundamentals are similar, but some things are out of date as PyCalphad has improved in various ways. 

To address what you've specifically asked about:
> This would involve starting with the information available in a TDB file, selecting the relevant models for the chosen problem, and building up the optimization problem by hand. Even better would be an example where the problem isn't trivial (i.e. predicting a binary phase diagram for an alloy with >4 components).
I can share my suggestions. 
1. TDB files: the closest thing to a specification for TDB files is the Database Manager's Guide in the [Thermo-Calc Documentation Set](https://thermocalc.com/support/documentation/). It's fairly exhaustive. This and just reading a lot of TDBs will take you a long way.
2. Selecting models: the reference I always go back to is Chapter 5 of Lukas, Fries, and Sundman "Computational Thermodynamics: The Calphad Method" 2007 https://doi.org/10.1017/CBO9780511804137. This covers most of the models relevant for alloys and many other materials. There are other models that exist, such as the modified quasi-chemical models, which I find myself going back to the recent paper by Poschmann *et al.* 2021 https://doi.org/10.1016/j.calphad.2021.102341 and the book by Pelton "Phase Diagrams and Thermodynamic Modeling of Solutions" 2019 https://doi.org/10.1016/C2013-0-19504-9. 
3. Building the optimization problem: some of this is discussed in the textbooks I mentioned. The set up of an optimization problem is covered at a relatively basic level in Section 2.3 of the Lukas/Fries/Sundman book. A more rigorous treatment with details on the minimization method (which PyCalphad has mostly adopted) is probably best described in the paper from Sundman, Lu, and Ohtani 2015 https://doi.org/10.1016/j.commatsci.2015.01.029. We recently revisited this in a paper focusing on derivates and a metric for phase stability (Kunselman *et al.* 2024 https://doi.org/10.1016/j.calphad.2024.102705). A different, but still state of the art numerical treatment is nicely written up by Piro *et al.* 2013 https://doi.org/10.1016/j.commatsci.2012.09.011, if you're interested to se another way to solve the same problem.
4. Computations in multicomponent systems: a binary (=2 components) in a system of >4 components is a little ambiguous to me. If you mean to do equilibrium calculations in general for multicomponent systems, I think the latter papers I mentioned in (3) are sufficiently general that the equations can be applied to multicomponent systems (this is how the codes work). If you mean specifically on computing a 2D temperature-composition phase diagram for a system with >4 components, the method PyCalphad uses to compute phase diagrams is closely related to the method described in another paper by Sundman, Dupin, and Hallstedt 2021 https://doi.org/10.1016/j.calphad.2021.102330. 

There's a lot of good individual resources out there. These are just some of the ones I have found useful learning all this stuff. I've certainly left many quality references ones out simply because they are not currently at the top of my mind.

Hopefully this is helpful. I'm happy to expand further on specific topics and would be interested to see what answers others in the community have as well.

**Reply by Jarrod-Angove** [reply]
Wow, this is an incredibly thorough answer! Thank you so much for taking the time to put this all together. The slides you provided are particularly useful. I also hadn't tried many of the texts you suggested, so this is an amazing starting point. 

One step closer to fully understanding the models I'm using in my thesis!
